Improve error message with type constructors with no parameters

```gleam
pub type ABC(a) {
  A(val: a)
}

pub type B(a) {
  B(ABC, ABC(a))
}
```
Produces this error:
```txt
error: Incorrect arity
  ┌─ /src/main.gleam:7:5
  │
6 │   B(ABC, ABC(a))
  │     ^^^ Expected 1 argument, got 0

Functions and constructors have to be called with their expected number of
arguments.
```
I think error message could be improved here:
```diff
 error: Incorrect arity
   ┌─ /src/main.gleam:7:5
   │
 6 │   B(ABC, ABC(a))
+  │     ^^^ Expected 1 type argument, got 0
-  │     ^^^ Expected 1 argument, got 0
 
+ `ABC` requires 1 type argument, but none were provided.
- Functions and constructors have to be called with their expected number of
- arguments.
```
Citing functions is a bit misleading here, I think internally in the compiler we call `ABC` a _type constructor_ but that term is never explained anywhere in the tour and we only ever use _constructor_ to refer to the variants of a type, not the type constructor itself; so that's a bit confusing as well. I think it would be simpler and less confusing if we just said that the type needs parameters that were not provided


For reference this is what Rust does:
```txt
error[E0107]: missing generics for struct `ABC`
  --> src/main.rs:13:21
   |
13 | struct B<A>(ABC<A>, ABC);
   |                     ^^^ expected 1 generic argument
   |
note: struct defined here, with 1 generic parameter: `A`
  --> src/main.rs:9:8
   |
9  | struct ABC<A> {
   |        ^^^ -
help: add missing generic argument
   |
13 | struct B<A>(ABC<A>, ABC<A>);
   |                        +++


Generic type 'ABC' requires 1 type argument(s).
```
I'm not sure about pointing to the type and suggesting to add one, but the wording of the error is nice, we could take inspiration from that!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -73,6 +73,31 @@
   errors as soon as it finds one.
   ([Giacomo Cavalieri](https://github.com/giacomocavalieri))
 
+- The error message for types used with the wrong number of arguments has been
+  improved. For example, this piece of code:
+
+  ```gleam
+  type Wibble(a)
+
+  type Wobble {
+    Wobble(Wibble)
+  }
+  ```
+
+  Produces the following error:
+
+  ```txt
+    error: Incorrect arity
+    ┌─ /src/one/two.gleam:5:10
+    │
+  5 │   Wobble(Wibble)
+    │          ^^^^^^ Expected 1 type argument, got 0
+
+  `Wibble` requires 1 type argument but none where provided.
+  ```
+
+  ([Giacomo Cavalieri](https://github.com/giacomocavalieri))
+
 ### Build tool
 
 - The build tool now supports placing modules in a directory called `dev`,
@@ -156,6 +181,7 @@
     }
   }
   ```
+
   ([Surya Rose](https://github.com/GearsDatapacks))
 
 - The language server now provides hover, autocomplete and goto definition

--- a/compiler-core/src/error.rs
+++ b/compiler-core/src/error.rs
@@ -2163,16 +2163,20 @@ But function expects:
                 TypeError::IncorrectTypeArity {
                     location,
                     expected,
-                    given,
-                    ..
+                    given: given_number,
+                    name,
                 } => {
-                    let text = wrap("Functions and constructors have to be \
-called with their expected number of arguments.");
                     let expected = match expected {
-                        0 => "no arguments".into(),
-                        1 => "1 argument".into(),
-                        _ => format!("{expected} arguments"),
+                        0 => "no type arguments".into(),
+                        1 => "1 type argument".into(),
+                        _ => format!("{expected} type arguments"),
                     };
+                    let given = match given_number {
+                        0 => "none",
+                        _ => &format!("{given_number}")
+                    };
+                    let text = wrap_format!("`{name}` requires {expected} \
+but {given} where provided.");
                     Diagnostic {
                         title: "Incorrect arity".into(),
                         text,
@@ -2180,7 +2184,7 @@ called with their expected number of arguments.");
                         level: Level::Error,
                         location: Some(Location {
                             label: Label {
-                                text: Some(format!("Expected {expected}, got {given}")),
+                                text: Some(format!("Expected {expected}, got {given_number}")),
                                 span: *location,
                             },
                             path: path.clone(),

--- a/compiler-core/src/type_/tests/errors.rs
+++ b/compiler-core/src/type_/tests/errors.rs
@@ -3071,3 +3071,16 @@ pub fn main() {
 "#
     );
 }
+
+#[test]
+fn error_for_missing_type_parameters() {
+    assert_module_error!(
+        r#"
+type Wibble(a)
+
+type Wobble {
+  Wobble(Wibble)
+}
+"#
+    );
+}

